phoenix-core/src/main/java/org/apache/phoenix/hbase/index/Indexer.java

```diff
@@ -506,30 +623,130 @@ public void preBatchMutateWithExceptions(ObserverContext<RegionCoprocessorEnviro
+                          if (!context.pendingRows.contains(rowKey)) {
+                              Put put = new Put(m.getRow());
+                              put.addColumn(emptyCF, emptyCQ, ts, TRUE_BYTES);
+                              postIndexUpdates.add(new Pair<>(new Pair<>(put, next.getFirst().getSecond()), next.getSecond()));
+                          }
+                      } else {
+                          // For a delete mutation, first unverify the exiting row in the index table and then delete
+                          // the row from the index table after deleting the corresponding row from the data table
+                          indexUpdatesItr.remove();
+                          Put put = new Put(m.getRow());
+                          put.addColumn(emptyCF, emptyCQ, ts, FALSE_BYTES);
+                          indexUpdatesForDeletes.add(new Pair<>(put, next.getFirst().getSecond()));
+                          // Ignore post index updates (i.e., the third write phase updates) for this row if it is
+                          // going through concurrent updates
+                          RowKey rowKey = new RowKey(next.getSecond());
+                          if (!context.pendingRows.contains(rowKey)) {
+                              postIndexUpdates.add(next);
+                          }
+                      }
+                  }
+              }
           }
           if (!localUpdates.isEmpty()) {
               miniBatchOp.addOperationsFromCP(0,
                   localUpdates.toArray(new Mutation[localUpdates.size()]));
           }
-          if (!indexUpdates.isEmpty()) {
-              context.indexUpdates = indexUpdates;
-              // write index updates to WAL
-              if (durability != Durability.SKIP_WAL) {
-                  // we have all the WAL durability, so we just update the WAL entry and move on
-                  for (Pair<Mutation, byte[]> entry : indexUpdates) {
-                    edit.add(new IndexedKeyValue(entry.getSecond(), entry.getFirst()));
-                  }              
+          if (!indexUpdatesForDeletes.isEmpty()) {
+              context.indexUpdates = indexUpdatesForDeletes;
+          }
+
+          if (!indexUpdates.isEmpty() && context.indexUpdates.isEmpty()) {
+              context.indexUpdates = new ArrayList<>(indexUpdates.size());
+          }
+          for (Pair<Pair<Mutation, byte[]>, byte[]> update : indexUpdates) {
+                  context.indexUpdates.add(update.getFirst());
+          }
+      }
+      // Sleep for one millisecond if we have prepared the index updates in less than 1 ms. The sleep is necessary to
+      // get different timestamps for concurrent batches that share common rows. It is very rare that the index updates
+      // can be prepared in less than one millisecond
+      if (!context.rowLocks.isEmpty() && now == EnvironmentEdgeManager.currentTimeMillis()) {
+          Thread.sleep(1);
+          LOG.debug("slept 1ms");
```

Comment:
Do we really need this log? I am not sure if it will help debugging. Perhaps adding index name makes it more useful?

Reply:
To make sure that it is a rare event. I will add the name of the index table.

Reply:
Done